### ISSUE TITLE
UTC functions

### ISSUE THREAD

**Issue by leshakoss** [date-fns/date-fns#376]
Make UTC versions of functions and make non-UTC versions (regular ones) be based on UTC-versons.

Example:

```javascript
// date-fns/utc/addDays

import toDate from '../toDate'

export default function addDays (dirtyDate, amount, options) {
  const date = toDate(dirtyDate, options)
  date.setUTCDate(date.getDate() + amount)
  return date
}
```

```javascript
// date-fns/utc/subDays

import addDays from '../addDays'

export default function subDays (dirtyDate, amount, options) {
  return addDays(dirtyDate, -amount, options)
}
```

```javascript
// date-fns/subDays

import utcSubDays from '../utc/subDays'
import toUTC from '../_lib/toUTC'
import fromUTC from '../_lib/fromUTC'

export default function subDays (dirtyDate, amount, options) {
  const utcDate = toUTC(dirtyDate)
  const utcResult = utcSubDays(utcDate, amount, options)
  return fromUTC(utcResult)
}
```

The problem with non-UTC functions is that they are not pure and can produce the different results in the different timezones. Moreover, in some timezones they can produce the wrong result. That happens when the date just doesn't exist in a timezone:

```
$ env TZ="America/New_York" node
> new Date(2017, 2 /* March */, 12, 2).toString()
'Sun Mar 12 2017 03:00:00 GMT-0400 (EDT)'
> new Date(2017, 2 /* March */, 12, 3).toString()
'Sun Mar 12 2017 03:00:00 GMT-0400 (EDT)'
```

In `America/New_York` timezone the date `2017-03-12T02:00:00` just doesn't exist because DST.
So when in chain of functions some intermediate result should be `2017-03-12T02:00:00`, the result of the chain will be wrong in `America/New_York` timezone. Sadly, it can also happen without chaining, because some of our functions can depend on the other functions. So every non-UTC function is not pure and produces side effects and subtle bugs. The way to resolve this issue is to make non-UTC operations minimal (is the best case, it should be done only twice: to convert local date to UTC and to convert the result back to the local timezone)

**Comment by jbrantly** [comment]
We just ran into a situation where we needed to `format` dates in UTC but were unable to using date-fns since it only formats in local time. I'm assuming that feature would be bundled into this issue? Very much looking forward to this!

**Comment by hlehmann** [comment]
Same issue with startOfDay in UTC

**Comment by hlehmann** [comment]
Is this coming with v2 ?

**Comment by leshakoss** [comment]
@hlehmann sadly, not in `v2.0.0`, but v2 features locales modified to work in UTC functions as well as in regular ones, so we plan to release UTC as `v2.x`

**Comment by hlehmann** [comment]
cool, and really nice job with the fp. Hope to see UTC, most of my data are in UTC and I still need to use moment for those.

**Comment by thisconnect** [comment]
+1

**Comment by drpotato** [comment]
Really look forward to UTC versions being added.

**Comment by hlehmann** [comment]
Could it be possible to have conversion functions like localeToUtc and utcToLocale, keeping the same hour but regarding the new timezone.

It would allow:
- to use date-fns until functions are ready
- be useful for example when components (like a date picker) which it's internal opérations use locale hours but we want to have it in utc.

**Comment by abritinthebay** [comment]
Question: does this feature have a suggested test rig around it? If so it should be pretty simple to produce the *base* functionality and then work on the exception (ie - mostly DST) handling.

**Comment by evenfrost** [comment]
In case it helps anyone, while it's WIP, I'm using helper function to convert date to UTC and then use it with `date-fns`:
```js
const { format: formatDate } = require('date-fns');

const getUTCDate = (dateString = Date.now()) => {
  const date = new Date(dateString);

  return new Date(
    date.getUTCFullYear(),
    date.getUTCMonth(),
    date.getUTCDate(),
    date.getUTCHours(),
    date.getUTCMinutes(),
    date.getUTCSeconds(),
  );
};

formatDate(new Date(), 'DD MMM, YYYY HH:mm:ss') // returns local time
formatDate(getUTCDate(), 'DD MMM, YYYY HH:mm:ss') // returns UTC time

**Comment by benmccann** [comment]
@leshakoss I'm wondering if the best solution is to define a UTC version of each function? It looks to me like this will affect almost every function in the library. And when we support additional timezones besides the user's and UTC we will need a third function for taking the timezone information:

```
// date-fns/subDays
export default function subDays (dirtyDate, amount, options) {
```

```
// date-fns/utc/subDays
export default function subDays (dirtyDate, amount, options) {
```

```
// date-fns/iana/subDays
export default function subDays (dirtyDate, tz, amount, options) {
```

Perhaps we should add the timezone as a field in the `options` object instead? Then we could keep just one function

**Comment by benmccann** [comment]
I've send a pull request for full IANA timezone support (including UTC): https://github.com/date-fns/date-fns/pull/676

**Comment by cbovis** [comment]
Is there a preferred approach for addressing this? Additional methods e.g. addDaysUTC or additional options to existing methods?

I'd love to start contributing on this one if there's an agreement on the API.

**Comment by leshakoss** [comment]
@cbovis I've made a PoC a while ago: https://github.com/date-fns/date-fns/pull/625
I'd say we wait for v2 beta to make sure that API won't change anymore before doing it

**Comment by fpintos** [comment]
I found myself in the same pickle; we have dates in different time zones and the only reliable way to represent them was to use the UTC fields and doing so requires UTC-based versions of the functions. I ended up writing the few ones I needed.

**Comment by LinBoLen** [comment]
I just need format can support utc.

**Comment by chawes13** [comment]
Similar to @evenfrost 's approach, we are using a utility in the meantime as a workaround that leverages `getTimezoneOffset`.

```js
const getUTCDate = (date = new Date()) => { 
  return new Date(date.getTime() + date.getTimezoneOffset()*60*1000)
}
```

**Comment by grissius** [comment]
@chawes13 However this does not solve the issue with [DST](https://en.wikipedia.org/wiki/Daylight_saving_time) :crying_cat_face: 

`getTimezoneOffset` is constant no matter the current date, but the locale hour changes according to DST, wheras in UTC it remains the same throughout the year.

**Comment by chawes13** [comment]
@grissius According to [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getTimezoneOffset#Description), `getTimezoneOffset` should change as well.

> The time zone offset returned is the one that applies for the Date that it's called on. Where the host system is configured for daylight saving, the offset will change depending on the date and time that the Date represents and that daylight saving applies.

```js
// For DST in the US (March 10, 2019 at 2:00 am)
const beforeDST = new Date('2019-03-09 6:')
const afterDST = new Date('2019-03-10 6:')

beforeDST.getTimezoneOffset() // 360
afterDST.getTimezoneOffset() // 300

getUTCDate(beforeDST) // Sat Mar 09 2019 12:00:00 GMT-0600 (Central Standard Time)
getUTCDate(afterDST) // Sun Mar 10 2019 11:00:00 GMT-0500 (Central Daylight Time)
```

**Comment by Rendez** [comment]
Based on the previous answers, here's my two cents to normalize different time zones to UTC:
```js
/**
 * This function assumes that an input with a time zone has to be treated as UTC.
 * This is helpful to correct dates stored in database with the wrong time zone.
 *
 * @param {string} dateStr
 * @returns {Date} date object with the adjustments from the target time zone
 */
export function utcDate(dateStr) {
    const ts = Date.parse(dateStr);
    if (!ts) {
        throw new Error('Invalid Date');
    }

    const date = new Date(ts);
    if (hasTimezone(dateStr)) {
        return new Date(date.getTime() + date.getTimezoneOffset() * 6e4);
    }

    return date;
}

const patterns = [
    /(Z)$/, // Z
    /([+-])(\d{2})$/, // ±hh
    /([+-])(\d{2}):?(\d{2})$/, // ±hh:mm or ±hhmm
    /(UTC|(?:[a-zA-Z]+\/[a-zA-Z_]+(?:\/[a-zA-Z_]+)?))$/, // IANA time zone
];

/**
 * Returns true if it matches any of the time zone offset formats
 *
 * @param {string} timezoneString
 * @returns {boolean}
 */
export function hasTimezone(timezoneString) {
    return patterns.some((pattern) => pattern.exec(timezoneString));
}
```

**Comment by kasperlauge** [comment]
@evenfrost

Doesn't your solution do the opposite of the intention? I just compared the result from your solution:

```
var h = new Date("2019-05-20T10:00");
var u = new Date(h.getUTCFullYear(),h.getUTCMonth(), h.getUTCDate(), h.getUTCHours(),h.getUTCMinutes(),h.getSeconds());

> u
> Mon May 20 2019 08:00:00 GMT+0200
```

Compared with moment:

```
var m = moment.utc("2019-05-20T10:00").toDate();
> m
> Mon May 20 2019 12:00:00 GMT+0200
```

Or maybe I misunderstood something?

**Comment by evenfrost** [comment]
@kasperlauge 

You pass a date to `moment.utc` and it converts it to UTC date (in your case it's 10:00 GMT). Then, you use `.toDate`, which outputs this date in your local timezone (e.g. 12:00 GMT+0200, means + 2 hours).

**Comment by mesterum** [comment]
@evenfrost nice hack but
`getUTCDate(new Date("2017-03-11T22:00")) -`
`getUTCDate(new Date("2017-03-11T21:00"))` is 0 for  
$ env TZ="America/New_York" node

> The problem with non-UTC functions is that they are not pure and can produce the different results in the different timezones.

**Comment by jayarjo** [comment]
Inability to properly and easily do UTC kills the whole hype around `date-fns` kinda :/ Do something.

@mesterum what about @chawes13 acpproach?

**Comment by jayarjo** [comment]
@leshakoss @benmccann I think instead of providing separate functions for UTC, you should simply operate on a wrapper of Date object. What's the point to expose that Date object everywhere if it's inadequate.

**Comment by kossnocorp** [comment]
@jayarjo what do you mean saying wrapper? Please provide code example of how you see it.

**Comment by jayarjo** [comment]
More or less the same Date object (with all the same methods and properties, following the same interface that is), but with ability to change the timezone. I can't say I've given a deep though to it, but I imagine something like [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) around the Date object should work.

**Comment by kossnocorp** [comment]
@jayarjo this would make build unreasonable massive. If you need such an API, there's Luxon: https://moment.github.io/luxon. I don't see a point of making a clone. date-fns focus is to help you to keep build size small and performant, and it will fail on that if we would make our own wrapper.

**Comment by jayarjo** [comment]
I could probably try to PR something.

**Comment by kossnocorp** [comment]
Ok, please do. If you can, please share a code snippet with the idea so that I can grasp it.

**Comment by bagaskarala** [comment]
i still confuse how to format `2019-10-16T10:10:54+08:00` to be `16-10-2019 10:10:54` in my local time(gmt+7)

i want to be formatted as is, not converted to my local, ho to do that?

note: i dont want to include IANA `Asia/Makassar` in formatting, i want just using that string only

**Comment by stevenvachon** [comment]
Even something like a `subTimezoneOffset()` would be nice.

**Comment by stevenvachon** [comment]
```js
const startOfTodayUTC = () => new Date(new Date().setUTCHours(0,0,0,0));
const endOfTodayUTC = () => new Date(new Date().setUTCHours(23,59,59,999));
```

**Comment by empperi** [comment]
I'm kinda surprised this is still open after almost 3 years but I'll add my opinion on this.

As far as I see it, there definitely should not be separate UTC functions for everything. Instead, each and every function in the whole library should accept timezone as parameter (via options) and that should be respected.  The `date-fns-tz` library doesn't really solve anything else than formatting issues.

For example, when writing tests one always wants to execute everything in UTC timezone. This is because when those tests are executed on a CI environment the chances are quite high that server is in a different timezone than what the developers are in - especially in today's cloud based CI services. 

The common way to tackle this is to set the `TZ` env variable to `UTC` but unfortunately this is not a real solution to anything since that hack [doesn't work for example on Windows at all](https://github.com/nodejs/node/issues/4230). The reason for this is that `TZ` environment variable isn't actually a V8 specific thing but it is a Unix thing. It's just pure luck that it actually works with V8 since V8 fetches the default timezone from the operating system and with unix the operating system respects the `TZ` env variable override.

So, when combined with the fact that `date-fns` relies on JavaScript Date objects it is inherently bound by it's quirks such as it's insistence on working on default timezones. These are inconsistent by nature and cannot be overridden in any reliable way for all operating systems.

Here is a very simple example of how things can very easily break. If this code is executed on a machine which is in GMT+3 timezone it produces a logically wrong result:

```javascript
isSameDay(new Date('2019-09-02T21:00:00.000Z'), new Date('2019-09-03T00:00:00.000Z'));
// returns true
```

Both dates are defined as UTC but the comparison is done based on default timezone, thus both dates are adjusted by +3 hours and suddenly they indeed are representing the same day. This gets even worse if this is executed in `Europe/Helsinki` timezone:

```javascript
isSameDay(new Date('2019-09-02T21:00:00.000Z'), new Date('2019-09-03T00:00:00.000Z'));
// returns true during summer, false in winter due to daylight saving
```

This is a text book example of a flaky and non-deterministic comparison which is horrible in tests.  In order to bypass this problem one needs to be constantly aware of it and write hacks and conversions everywhere - stuff I'd prefer a dedicated date library to do for me instead. Using `date-fns-tz` library [time zone helper](https://www.npmjs.com/package/date-fns-tz#time-zone-helpers) functions doesn't help at all since those still return regular js Date objects which use the default timezone under the hood.

Basically the only correct way to implement this would be the following:

```javascript
isSameDay(new Date('2019-09-02T21:00:00.000Z'), 
          new Date('2019-09-03T00:00:00.000Z'), 
          { timeZone: 'UTC' });
```

which then under the hood would ensure all Date objects being compared are first adjusted to the given timezone and only then do the comparison. This change would be backwards compatible (if no timeZone is given then use default timezone) and would provide all the flexibility one could wish for when operating with timezones.

**Comment by kossnocorp** [comment]
@empperi I appreciate the input but what you propose will significantly increase the build size and that's definitely what we want. We're looking into an alternative solution with inheriting from `Date` but it's early in development.

**Comment by empperi** [comment]
> @empperi I appreciate the input but what you propose will significantly increase the build size and that's definitely what we want. We're looking into an alternative solution with inheriting from `Date` but it's early in development.

I understand although I got to say I would highly prefer a consistently working library over minimal footprint. Hope you'll get this sorted out quickly since this is a major issue in `date-fns`.

**Comment by kossnocorp** [comment]
> I understand although I got to say I would highly prefer a consistently working library over minimal footprint. Hope you'll get this sorted out quickly since this is a major issue in date-fns.

@empperi working within local timezone covers the overwhelming majority of the use cases, so the minimal footprint will always be the priority over the convenience of working with time zones. I'm aware that for the ones who do need timezones, it's a significant issue, so that's still a priority for us.

**Comment by thomasnunez** [comment]
You can install momentjs...
Convert the string representation of the date to an object and it should work.
i.e. "selected={moment(createDate).toDate()}"

**Comment by zacheryph** [comment]
I am pretty sure this belongs here. I have noticed this lead to very basic issues regarding formatting. This is probably well known. This shows completely wrong values within `@material-ui/pickers` as well as when I search for a `type: "date"` field within `material-table` I will get results for the day previous of the date i entered.

An issue this simple is kind of frightening to exist this long actually.

> System Timezone: CST

```js
format(new Date('2019-12-16'), 'yyyy-MM-dd')
# => '2019-12-15'
```

**Comment by kossnocorp** [comment]
@zacheryph thanks for reporting, but that's a JavaScript quirk. When you parse the date string, it converts the date to UTC, but that's unrelated to this issue, which is not actually an issue but a feature request.

To parse a date in a local time zone I recommend to use `parseISO`:

```js
new Date('2019-11-16').toString()
//=> 'Sat Nov 16 2019 08:00:00 GMT+0800 (Malaysia Time)'

new Date(2019, 11, 16).toString()
//=> 'Mon Dec 16 2019 00:00:00 GMT+0800 (Malaysia Time)'

parseISO('2019-12-16').toString()
//=> 'Mon Dec 16 2019 00:00:00 GMT+0800 (Malaysia Time)'
```

**Comment by kossnocorp** [comment]
I appreciate everyone's input, but to avoid unrelated discussions, I'm locking the issue. If you want to share your use case, opinion about how it should work or whatever, please do it in the community chat: https://spectrum.chat/date-fns

**Comment by kossnocorp** [comment]
Done via https://github.com/date-fns/utc
